File: platform/cli-go/internal/commands/log.go
package commands

import (
	"fmt"
	"sort"
	"strconv"
	"time"

	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

type logEvent struct {
	Timestamp string
	Message   string
}

func formatLogTime(dateStr string) string {
	t, err := time.Parse(time.RFC3339, dateStr)
	if err != nil {
		return ""
	}
	return fmt.Sprintf("%02d:%02d", t.Hour(), t.Minute())
}

func formatDateHeader(dateStr string) string {
	t, err := time.Parse(time.RFC3339, dateStr)
	if err != nil {
		return dateStr
	}
	now := time.Now()

	tDay := time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
	nDay := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
	diffDays := int(nDay.Sub(tDay).Hours() / 24)

	if diffDays == 0 {
		return "Today"
	}
	if diffDays == 1 {
		return "Yesterday"
	}
	if diffDays < 7 {
		return fmt.Sprintf("%d days ago", diffDays)
	}
	return t.Format("Monday, Jan 2")
}

func getDateKey(dateStr string) string {
	t, err := time.Parse(time.RFC3339, dateStr)
	if err != nil {
		return dateStr
	}
	return fmt.Sprintf("%d-%02d-%02d", t.Year(), t.Month(), t.Day())
}

func newLogCmd() *cobra.Command {
	var limit string

	cmd := &cobra.Command{
		Use:   "log",
		Short: "Show activity log",
		RunE: func(cmd *cobra.Command, args []string) error {
			maxEntries, _ := strconv.Atoi(limit)
			if maxEntries <= 0 {
				maxEntries = 20
			}

			var events []logEvent

			// Collect events from features
			features := workspace.ListFeatures()
			for _, feature := range features {
				events = append(events, logEvent{
					Timestamp: feature.Created,
					Message:   fmt.Sprintf("Started feature: %s", feature.Name),
				})

				// Collect from diffs
				diffs := workspace.LoadDiffsForFeature(feature)
				for _, d := range diffs {
					events = append(events, logEvent{
						Timestamp: d.Created,
						Message:   fmt.Sprintf("Created diff %s: %s", format.DiffLabel(d.Position), d.Title),
					})

					if d.Status == "landed" {
						events = append(events, logEvent{
							Timestamp: d.Updated,
							Message:   fmt.Sprintf("Landed: %s", d.Title),
						})
					} else if d.Status == "submitted" {
						events = append(events, logEvent{
							Timestamp: d.Updated,
							Message:   fmt.Sprintf("Submitted %s: %s", format.DiffLabel(d.Position), d.Title),
						})
					}
				}
			}

			// Collect from undo entries
			undoEntries := workspace.ListUndoEntries()
			for _, entry := range undoEntries {
				if entry.Action == "sync" {
					events = append(events, logEvent{
						Timestamp: entry.Timestamp,
						Message:   fmt.Sprintf("Synced: %s", entry.Description),
					})
				}
			}

			// Sort by timestamp descending
			sort.Slice(events, func(i, j int) bool {
				ti, _ := time.Parse(time.RFC3339, events[i].Timestamp)
				tj, _ := time.Parse(time.RFC3339, events[j].Timestamp)
				return ti.After(tj)
			})

			// Deduplicate and limit
			seen := make(map[string]bool)
			var uniqueEvents []logEvent
			for _, event := range events {
				key := event.Timestamp + ":" + event.Message
				if !seen[key] {
					seen[key] = true
					uniqueEvents = append(uniqueEvents, event)
				}
				if len(uniqueEvents) >= maxEntries {
					break
				}
			}

			if len(uniqueEvents) == 0 {
				fmt.Println(format.DimText("No activity yet."))
				return nil
			}

			// Group by day (preserve order)
			type dayGroup struct {
				Key    string
				Events []logEvent
			}
			var groups []dayGroup
			groupMap := make(map[string]int)

			for _, event := range uniqueEvents {
				key := getDateKey(event.Timestamp)
				if idx, ok := groupMap[key]; ok {
					groups[idx].Events = append(groups[idx].Events, event)
				} else {
					groupMap[key] = len(groups)
					groups = append(groups, dayGroup{Key: key, Events: []logEvent{event}})
				}
			}

			for _, group := range groups {
				header := formatDateHeader(group.Events[0].Timestamp)
				fmt.Println(header)

				for _, event := range group.Events {
					timeStr := formatLogTime(event.Timestamp)
					fmt.Printf("  %s  %s\n", format.DimText(timeStr), event.Message)
				}
				fmt.Println()
			}

			return nil
		},
	}

	cmd.Flags().StringVarP(&limit, "limit", "n", "20", "Number of entries to show")

	return cmd
}

File: platform/cli/src/commands/log.ts
import { Command } from 'commander';
import chalk from 'chalk';
import {
  listFeatures,
  loadDiffsForFeature,
  listUndoEntries,
} from '../workspace';
import { errorMessage, diffLabel } from '../format';

interface LogEvent {
  timestamp: string;
  message: string;
}

function formatTime(date: string): string {
  const d = new Date(date);
  const hours = d.getHours().toString().padStart(2, '0');
  const minutes = d.getMinutes().toString().padStart(2, '0');
  return `${hours}:${minutes}`;
}

function formatDateHeader(date: string): string {
  const d = new Date(date);
  const now = new Date();

  const dDay = new Date(d.getFullYear(), d.getMonth(), d.getDate());
  const nDay = new Date(now.getFullYear(), now.getMonth(), now.getDate());
  const diffDays = Math.floor(
    (nDay.getTime() - dDay.getTime()) / (1000 * 60 * 60 * 24),
  );

  if (diffDays === 0) return 'Today';
  if (diffDays === 1) return 'Yesterday';
  if (diffDays < 7) return `${diffDays} days ago`;
  return d.toLocaleDateString('en-US', {
    weekday: 'long',
    month: 'short',
    day: 'numeric',
  });
}

function getDateKey(date: string): string {
  const d = new Date(date);
  return `${d.getFullYear()}-${(d.getMonth() + 1).toString().padStart(2, '0')}-${d.getDate().toString().padStart(2, '0')}`;
}

export function registerLogCommand(program: Command): void {
  program
    .command('log')
    .description('Show activity log')
    .option('-n, --limit <n>', 'Number of entries to show', '20')
    .action(async (opts: { limit: string }) => {
      try {
        const limit = parseInt(opts.limit, 10) || 20;
        const events: LogEvent[] = [];

        // Collect events from features
        const features = listFeatures();
        for (const feature of features) {
          events.push({
            timestamp: feature.created,
            message: `Started feature: ${feature.name}`,
          });

          // Collect from diffs
          const diffs = loadDiffsForFeature(feature);
          for (const d of diffs) {
            events.push({
              timestamp: d.created,
              message: `Created diff ${diffLabel(d.position)}: ${d.title}`,
            });

            if (d.status === 'landed') {
              events.push({
                timestamp: d.updated,
                message: `Landed: ${d.title}`,
              });
            } else if (d.status === 'submitted') {
              events.push({
                timestamp: d.updated,
                message: `Submitted ${diffLabel(d.position)}: ${d.title}`,
              });
            }
          }
        }

        // Collect from undo entries
        const undoEntries = listUndoEntries();
        for (const entry of undoEntries) {
          if (entry.action === 'sync') {
            events.push({
              timestamp: entry.timestamp,
              message: `Synced: ${entry.description}`,
            });
          }
        }

        // Sort by timestamp descending
        events.sort(
          (a, b) =>
            new Date(b.timestamp).getTime() -
            new Date(a.timestamp).getTime(),
        );

        // Deduplicate and limit
        const seen = new Set<string>();
        const uniqueEvents: LogEvent[] = [];
        for (const event of events) {
          const key = `${event.timestamp}:${event.message}`;
          if (!seen.has(key)) {
            seen.add(key);
            uniqueEvents.push(event);
          }
          if (uniqueEvents.length >= limit) break;
        }

        if (uniqueEvents.length === 0) {
          console.log(chalk.dim('No activity yet.'));
          return;
        }

        // Group by day
        const groups = new Map<string, LogEvent[]>();
        for (const event of uniqueEvents) {
          const key = getDateKey(event.timestamp);
          if (!groups.has(key)) {
            groups.set(key, []);
          }
          groups.get(key)!.push(event);
        }

        for (const [dateKey, dayEvents] of groups) {
          const header = formatDateHeader(dayEvents[0]!.timestamp);
          console.log(chalk.bold(header));

          for (const event of dayEvents) {
            const time = formatTime(event.timestamp);
            console.log(
              `  ${chalk.dim(time)}  ${event.message}`,
            );
          }
          console.log();
        }
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}
